src/fs/protocol/private/mmpt.ts

```diff
@@ -47,43 +48,53 @@ export default class MMPT implements Puttable {
   }
 
   async add(name: string, value: CID): Promise<void> {
-    if(!isNibble(name[0])) {
+    if (!isNibble(name[0])) {
       throw new Error(`Not a valid name, must be hexadecimal`)
     }
     const nextNameOrSib = this.nextTreeOrSiblingName(name)
 
     // if already in tree, then skip
-    if(name === nextNameOrSib){
+    if (name === nextNameOrSib) {
       // skip
     }
 
     // if no children starting with first char of name, then add with entire name as key
-    else if(nextNameOrSib === null) {
+    else if (nextNameOrSib === null) {
       this.links[name] = link.make(name, value, true, 0)
     }
 
     // if multiple children with first char of names, then add to that tree
-    else if(nextNameOrSib.length === 1){
+    else if (nextNameOrSib.length === 1) {
       const nextTree = await this.getDirectChild(nextNameOrSib)
       await nextTree.add(name.slice(1), value)
-      await this.putAndUpdateLink(nextNameOrSib, nextTree)
+      await this.putAndUpdateChildLink(nextNameOrSib)
     }
 
     // if one other child with first char of name, then put both into a child tree
-    else{
+    else {
       const newTree = this.addEmptyChild(name[0])
       const nextCID = this.links[nextNameOrSib].cid
       this.removeChild(nextNameOrSib)
       await Promise.all([
         newTree.add(name.slice(1), value),
         newTree.add(nextNameOrSib.slice(1), nextCID)
       ])
-      await this.putAndUpdateLink(name[0], newTree)
+      await this.putAndUpdateChildLink(name[0])
     }
   }
 
-  async putAndUpdateLink(name: string, child: MMPT): Promise<void> {
-    const { cid, size } = await child.putDetailed()
+  async putAndUpdateChildLink(name: string): Promise<void> {
+    const cidBefore = this.links[name]?.cid
+    const { cid, size } = await this.children[name].putDetailed()
+    const cidNow = this.links[name]?.cid
+
+    if (cidBefore != cidNow) {
+      // If there are changes in-between, we have to
+      // re-try updating. Otherwise we can overwrite changes that
+      // a concurrent update made
+      return await this.putAndUpdateChildLink(name)
+    }
```

Comment:
Trying to understand the MMPT issue.
If the link has updated while updating it, what does that mean exactly?

Reply:
Ok, here's the best explanation I can give:

This is a race condition, so there's something going wrong when we have two entities concurrently writing to the MMPT. That's possible in javascript due to async-await: As soon as one "execution branch" (not sure how to call this) `await`s some system resource (e.g. network IO or indexedDB IO due to js-ipfs), another branch will continue its execution.
In rare cases, it's possible that two branches execute the same code path, but with slightly different data.

This means that it's possible that two `MMPT.add` calls try to update the same child tree, but try to do that with different data, because they were called for different reasons (different parameters in `MMPT.add()`).

Now, what *can* happen, is that both execute this line:
```js
const { cid, size } = await child.putDetailed()
```
And once the first one enters the `await`, the second one hits the same `await` call. And after the first one finishes, it'll run this line:
```js
this.links[name] = link.make(name, cid, false, size)
```
(and continue with something else until the next `await`)

And then the second branch *also* set's the same link with the same name to another CID. However, since both of these calls happened concurrently, the second time the link was set, it didn't include changes from the first time the link was set, thus dropping one or more branches from the actual DAG that'll be stored in IPFS.

---

The fix for this is to detect a race and the loser of the race will have to run again.

Also, note that in the fix we're using `this.children[name]` instead of the `child` parameter.
This is because `this.children[name]` will always give us the most recent *actual* child of the MMPT. The MMPT actually exists both in IPFS as a tree of CIDs *and* partially as a cached in-memory tree structure with the entries to `this.children` and `this.link` respectively.
During testing, Daniel and I noticed that the in-memory version would never drop anything that was `.add()`ed, but what's behind the `link`s *would*. That's because the in-memory version will always be modified synchronously.

---

Hope this clarifies the issue a bit :)

Reply:
Awesome, thanks!

> This means that it's possible that two MMPT.add calls try to update the same child tree, but try to do that with different data, because they were called for different reasons (different parameters in MMPT.add()).

I was primarily thinking around this, why it would update the same child tree. But I guess that's unavoidable? (I don't entirely remember how the MMPT works in relation to the other pieces). Anyhow, your solution works out perfectly 👍

Reply:
Well, the MMPT is append-only, and the keys used are SHA256 hex-encoded hashes, so it's also super unlikely that two things try to update the same child tree intuitively at first.

However, you have to remember that the IPFS DAG that encodes the tree is immutable. So anytime, say, a 4th-level child MMPT gets another node added, it's parent and grandparent and grand-grand-parent have to be updated, since all the CIDs change.

That's when two calls can suddenly clash and race each other.

Also, interestingly, note here that this is only true for the `this.links` property, but not for `this.children`.